src/main/java/org/breedinginsight/services/TraitService.java

```diff
@@ -287,4 +232,163 @@ public void preprocessTraits(List<Trait> traits) {
+
+        Optional<Program> optionalProgram = programService.getById(programId);
+        if (!optionalProgram.isPresent()) {
+            throw new DoesNotExistException("Program does not exist");
+        }
+        Program program = optionalProgram.get();
+
+        Optional<User> optionalUser = userService.getById(actingUser.getId());
+        if (!optionalUser.isPresent()){
+            throw new InternalServerException("Could not find user in system");
+        }
+        User user = optionalUser.get();
+
+        // Retrieve all of our traits
+        List<TraitEntity> existingTraitEntities = new ArrayList<>();
+        ValidationErrors missingTraitValidationErrors = new ValidationErrors();
+        for (int i = 0; i < traits.size(); i++) {
+            TraitEntity existingTrait = traitDAO.fetchOneById(traits.get(i).getId());
+            if (existingTrait != null) {
+                existingTraitEntities.add(existingTrait);
+            } else {
+                missingTraitValidationErrors.addError(i, traitValidatorError.getTraitIdDoesNotExistMsg());
+            }
+        }
+
+        try {
+            checkAndPrepareTraits(programId, traits, true);
+        } catch (ValidatorException e) {
+            e.getErrors().merge(missingTraitValidationErrors);
+            throw e;
+        }
+
+        if (missingTraitValidationErrors.hasErrors()) {
+            throw new ValidatorException(missingTraitValidationErrors);
+        }
+
+        // Create the traits
+        List<Trait> updatedTraits = new ArrayList<>();
+
+        if (traits.size() > 0) {
+
+            updatedTraits = dsl.transactionResult(configuration -> {
+
+                // Create the new observation levels
+                Map<String, ProgramObservationLevel> createdLevelsMap = checkAndCreateObservationLevels(programId, traits, actingUser);
+                List<Trait> updatedResults = new ArrayList<>();
+                for (int i = 0; i < traits.size(); i++) {
+
+                    Trait updatedTrait = traits.get(i);
+                    TraitEntity existingTraitEntity = existingTraitEntities.get(i);
+
+                    // Check observation level
+                    if (updatedTrait.getProgramObservationLevel().getId() == null) {
+                        ProgramObservationLevel level = createdLevelsMap.get(updatedTrait.getProgramObservationLevel().getName());
+                        updatedTrait.setProgramObservationLevel(level);
+                    }
+
+                    // Jump to scale
+                    ScaleEntity existingScaleEntity = scaleDAO.fetchOneById(existingTraitEntity.getScaleId());
+                    existingScaleEntity.setScaleName(updatedTrait.getScale().getScaleName());
+                    existingScaleEntity.setDataType(updatedTrait.getScale().getDataType());
+                    existingScaleEntity.setUpdatedBy(user.getId());
+                    scaleDAO.update(existingScaleEntity);
+
+                    // Update trait
```

Comment:
We also need to set active right?

Reply:
Wouldn't that be in a separate archive endpoint?

Reply:
based on later discussions, yes